_blog/2024-02-07-deploy-rust-web.mdx

```diff
@@ -0,0 +1,89 @@
+---
+title: Deploying a Rust web app
+description: This article explores the different ways to deploy Rust web services and the (dis)advantages of each. 
+author: josh
+tags: [rust, deployment, guide]
+thumb: write-rest-api-rust-thumb.png 
+cover: write-rest-api-rust-thumb.png 
+date: '2024-01-31T14:30:00'
+---
+
+This article will talk about the best way to deploy Rust. We’ll be covering all the ways that you can deploy Rust, including to VM/VPS, Serverless and options that are more in the middle.
+
+## How do you deploy Rust?
+
+When you compile Rust, it gets compiled to an executable. You can then run the binary from anywhere - including your own self-hosted server! 
+
+The compiled nature of Rust programs means that typically, the way to deploy them is through containerization or hosting them on a VPS. You compile the program into a Docker container (and add it to the machine if you're using a VPS), set up your proxy and SSL certifications if required, host it somewhere and it’s good to go. You can also use a web service that containerizes your application for you and then deploys it to their own network, setting up SSL/proxying automatically for you.
+
+## Different types of hosting
+
+### VPS
+
+VPS (Virtual Private Servers) are a way to be able to deploy to a machine that you have full control over and can  Typically when hosting via VPS, you have full control over everything in the VPS. You are able to SSH into the machine itself and set everything that you need up - for example, if you need some custom dependencies for your machine, you can do that. 
+
+The basic process involves mostly setting up Nginx or Apache (or a similar proxy), then grabbing the files you need from GitHub or GitLab and compiling the program itself. You can then set the web service up as a systemd service and it will automatically start up whenever the machine starts!
+
+Unfortunately, there are a couple of issues that come along with this: you need to handle all issues yourself (unless the VPS itself has an outage), and typically it’s less scalable than other deployment options. VPS machines typically have a given hard maximum for resource usage and you pay for a set amount - so if you’re not making full usage of the given resources, you may be overpaying. This is especially relevant to Rust, as Rust web applications are typically quite low memory footprint. 
+
+Here’s a list of advantages and disadvantages:
+
+| Advantage | Disadvantage |
+| --- | --- |
+| Total control over all aspects of the machine | If anything goes wrong you need to fix it yourself |
+| You know what you’re getting | Less scalable than serverless - machines typically have a hard limit |
+| Once you learn how to do it, it’s not too hard to do it again | You need to set everything up (Nginx, HTTPS, etc…) |
+| You can run whatever you want (NixOS, Ubuntu, Red Hat Enterprise Linux, etc…) | More expensive than other forms of deployment |
```

Comment:
:joy: NixOs, nice!! :joy:

Reply:
I would say not being reproducible is another disadvantage.

Reply:
Ie. if you have to start fresh because, of let's say an emergency, then you have to be able to recall all the manual setup you did last time to get it into a working state. The only way to get around this is to invest time in IaC setup.